### PR TITLE
#1141 Make more DynaFlow models Open Source

### PR DESCRIPTION
**Pull request by cossartque**

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model VRRemote "Model for centralized remote voltage regulation"
+
+import Dynawo.Connectors;
+import Dynawo.Types;
+
+  parameter Types.VoltageModulePu U0Pu "Start value of the regulated voltage in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:20]: U0Pu should be protected and appear after parameters and public variables (coding style ;) )

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  parameter Real Gain "Control gain";
+  parameter Types.Time tIntegral "Time integration constant";
+
+  Connectors.ZPin URefPu (value(start = U0Pu)) "Voltage regulation set point in p.u (base UNom)";
+  Connectors.ImPin URegulatedPu (value(start = U0Pu)) "Regulated voltage in p.u (base UNom)";
+  Connectors.ImPin NQ "Signal to change the reactive power generation of the generators participating in the centralized distant voltage regulation";
+
+  Types.PerUnit UErrorIntegralPu (start = 0) "Time-integral of the control error in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:28]: protected also I think

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+model VRRemote "Model for centralized remote voltage regulation"
+
+import Dynawo.Connectors;
+import Dynawo.Types;
+
+  parameter Types.VoltageModulePu U0Pu "Start value of the regulated voltage in p.u (base UNom)";
+  parameter Real Gain "Control gain";
+  parameter Types.Time tIntegral "Time integration constant";
+
+  Connectors.ZPin URefPu (value(start = U0Pu)) "Voltage regulation set point in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:24]: Better ImPin with a discrete tag in the extvar? I think we already had this discussion and don't exactly remember what we finally did...^^

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPQProp.mo [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model GeneratorPQProp "Model for generator PQ based on SignalN for the frequency handling and with a proportional reactive power regulation"
+
+ import Dynawo.Electrical.Machines;
+ import Dynawo.AdditionalIcons;
+ import Dynawo.Connectors;
+
+  extends Machines.BaseClasses.BaseGeneratorSimplified;
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPQProp.mo:21]: Could you extend Machines.SignalN.BaseClasses.BaseGeneratorSignalN?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model GeneratorPVDiagramPQ "Model for generator PV based on SignalN for the frequency handling with an N points PQ diagram."
+
+  import Dynawo.Connectors;
+  import Dynawo.Electrical.Machines;
+  import Modelica;
+  extends Machines.BaseClasses.BaseGeneratorSimplified;
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:20]: Could you extend Machines.SignalN.BaseClasses.BaseGeneratorSignalN?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/package.order [hunk 6]
```diff
@@ -1,4 +1,8 @@
 BaseClasses
+GeneratorPQProp
 GeneratorPV_INIT
 GeneratorPV
+GeneratorPVDiagram_INIT
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/package.order:5]: INIT model after dynamic model

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/CMakeLists.txt [hunk 7]
```diff
@@ -20,6 +20,8 @@ set(MODEL_FILES
     VRProportionalIntegral_INIT.mo
     VRProportionalReactiveFeedback.extvar
     VRProportionalReactiveFeedback.mo
+    VRRemote.extvar
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/CMakeLists.txt:23]: I am sorry I thought it was a machine regulation when I suggested this place in the library. Given the fact that it is a centralized  control (more as the SVC) it is better to put it in Controls>Voltage as you suggested first

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source time domain simulation tool for power systems.
+*/
+
+model GeneratorPVRemote "Model for generator PV based on SignalN for the frequency handling and a remote voltage regulation"
+
+  import Dynawo.Electrical.Machines;
+  extends Machines.BaseClasses.BaseGeneratorSimplified;
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:18]: Could you extend Machines.SignalN.BaseClasses.BaseGeneratorSignalN?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  parameter Types.ReactivePowerPu QMaxPu  "Maximum reactive power in p.u (base SnRef)";
+  parameter Types.PerUnit KGover "Mechanical power sensitivity to frequency";
+  parameter Types.ActivePower PNom "Nominal power in MW";
+  final parameter Real Alpha = PNom * KGover "Participation of the considered generator in the frequency regulation";
+
+  input Real URegulatedPu "Regulated voltage in p.u (base UNom)";
+  Connectors.ImPin N "Signal to change the active power reference setpoint of all the generators in the system in p.u (base SnRef)";
+  Connectors.ZPin alpha "Participation of the considered generator in the frequency regulation. It is equal to Alpha if the generator is not blocked, 0 otherwise.";
+  Connectors.ZPin alphaSum "Sum of all the participations of all generators in the frequency regulation";
+  Connectors.ZPin URefPu (value(start = URef0Pu)) "Voltage regulation set point in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:40]: ImPin?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  extends Machines.BaseClasses.BaseGeneratorSimplified;
+  extends AdditionalIcons.Machine;
+
+  type QStatus = enumeration (Standard "Reactive power is fixed to its initial value",
+                              AbsorptionMax "Reactive power is fixed to its absorption limit",
+                              GenerationMax "Reactive power is fixed to its generation limit");
+
+public
+
+  parameter Types.VoltageModulePu URef0Pu "Start value of the voltage regulation set point in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:27]: protected

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    Types.ReactivePowerPu QMaxPu(start = QMax0Pu) "Maximum reactive power in p.u (base SnRef)";
+
+  protected
+
+    Types.ActivePowerPu PGenRawPu (start = PGen0Pu) "Active power generation without taking limits into account in p.u (base SnRef) (generator convention)";
+
+  equation
+
+    PGenPu = tableQMin.u[1];
+    tFilter * der(QMinPu) + QMinPu = tableQMin.y[1];
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: I must admit that I don't like the use of QMin0Pu with the 0 in it (same in INIT model). Could you explain why did you need the variables QMinPu and QMaxPu with a dynamic evolution? Otherwise we could have had only QMinPu, QMaxPu as parameters and use tableQMin.y as final limit and same for tableQMax.y. Or would you have another suggestion/idea?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    parameter Types.Time tFilter "Filter time constant to update QMin/QMax";
+    parameter String QMinTableName "Name of the table in the text file to get QMinPu from PGenPu";
+    parameter String QMaxTableName "Name of the table in the text file to get QMaxPu from PGenPu";
+    parameter String QMinTableFile "Text file that contains the table to get QMinPu from PGenPu";
+    parameter String QMaxTableFile "Text file that contains the table to get QMaxPu from PGenPu";
+
+    Connectors.ImPin N "Signal to change the active power reference setpoint of all the generators in the system in p.u (base SnRef)";
+    Connectors.ZPin alpha "Participation of the considered generator in the frequency regulation";
+    Connectors.ZPin alphaSum "Sum of all the participations of all generators in the frequency regulation";
+    Connectors.ZPin URefPu (value(start = U0Pu)) "Voltage regulation set point in p.u (base UNom)";
```
Comment (MarianneSaugierRTE) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:39]: ImPin?

**Review by MarianneSaugierRTE — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo (hunk 1 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:20]: I'm note sure if I can put it as a protected parameter as it is directly used in the .par. What do you think ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo (hunk 2 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:28]: OK thanks

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo (hunk 3 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:24]: Yes we had the discussion. From what I remember and by looking at the existing models, I think we kept ZPin for the set points.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/package.order (hunk 6 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/package.order:5]: Ok thanks

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/CMakeLists.txt (hunk 7 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/CMakeLists.txt:23]: Ok I'll change it, no pb!

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo (hunk 1 above)
Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:20]: Hm... In fact as you don't use an init model here I don't know...

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo (hunk 9 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:40]: Same as before, we chose ZPin for the other models so maybe we keep ZPin here too ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo (hunk 12 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:39]: Same

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo (hunk 9 above)
Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:40]: The thing is: when you'll want to connect it to a continuous control like SVC+Rpcl for example you will be blocked. If the solution ImPin + discrete tag in the extvar works I would recommend that, what do you think?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo (hunk 11 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: Well if I put QMinPu and QMaxPu as parameters, and use tableQMin.y[1] and tableQMax.y[1] in the model as the actual QMinPu and QMaxPU (they change because they depend on PGenPu), I think it is quite confusing. My opinion is that we leave the impression that QMinPu and QMaxPu are fixed while they are in fact variables that depend on PGenPu. With the 0 I just give the Init value of QMinPu and QMaxPu, and they are then calculated in the model for each time step.
I could simply write QMinPu = tableQMin.y[1] (without the first order filter) but I have numerical issues. This tFilter parameter will use a default value and won't be chosen by the user. What do you think ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo (hunk 10 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:27]: Same as before, as I give this parameter in the .par, can it be protected ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo (hunk 3 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/VoltageRegulators/VRRemote.mo:24]: Ok I'll do it !

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo (hunk 9 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVRemote.mo:40]: Ok I'll do it then ! Thanks

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo (hunk 11 above)
Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: Ok then keep the dynamic ;)
But If the limits depend on PGenPu, how can you set the QMin0Pu and QMax0Pu to the right value without going through the tables?

Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: Good question! In fact they only need to respect QMin0Pu<Q0Pu<QMax0Pu, so I choose the largest value of the table (then it is replaced with the actual value at the first time step (maybe global init)). If I don't put an init value, it chooses 0, so I have QMinPu and QMaxPu that start at 0 with Q<>0, and it fails.

Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: Ok, in that case, could we have something like:
Types.ReactivePowerPu QMinPu(start = tableQMin[X])
in order not to have QMin0Pu in the par file (as it is not a real value...)?

Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: I'll try something like that and let you know

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPQProp.mo (hunk 4 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPQProp.mo:21]: I could yes. I will modify BaseGeneratorSignalN to have only the things that are common to all generators.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo (hunk 11 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: I doesn't seem to work this way...
It could work if the table was defined in the .mo but it is read in the .txt file so I cannot easily access it through Modelica.

Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo:52]: Ok thanks for trying

**Review by MarianneSaugierRTE — APPROVED** [review]
